src/compiler/parser.ts

```diff
@@ -1858,7 +1861,36 @@ module ts {
         function parseParameterInitializer() {
             return parseInitializer(/*inParameter*/ true);
         }
-
+        
+        function nextTokenIsIsKeyword(): boolean {
+            nextToken();
+            return token === SyntaxKind.IsKeyword;
+        }
+        
+        function parseTypeGuardType(parameterList?: NodeArray<ParameterDeclaration>): TypeGuardTypeNode {
+            let node = <TypeGuardTypeNode>createNode(SyntaxKind.TypeGuardType);
+            if (token === SyntaxKind.Identifier) {
+                node.target = createIdentifier(true);
+            }
+            else if (token === SyntaxKind.ThisKeyword) {
+                nextToken();
+            }
+            if (parseExpected(SyntaxKind.IsKeyword)) {
+                if (token === SyntaxKind.OpenBraceToken) {

```

Comment:
Is this even correct behavior? If I wrote
```ts
function hasAnonymousReturn(x): x is { prop: string } {
}
```
This would be an error. But shouldn't it be allowed?